fix issue #206 by updating the way exercises are stored and splitted

aricodeine opened the pull request:
This PR fixes a bug with the way the exercises under a plan are split and displayed on the start plan page.

fixes https://github.com/brandonp2412/Flexify/issues/206

Preview:
1) Exercise:
<img width="530" height="460" alt="image" src="https://github.com/user-attachments/assets/59b2b698-eadf-4173-9c2a-316a0b7ef55a" />

2) Plan preview:
<img width="556" height="679" alt="image" src="https://github.com/user-attachments/assets/bbca3f3d-e703-4df1-b7f7-5afce06d6d59" />

brandonp2412 commented:
Please include a [database migration](https://github.com/brandonp2412/Flexify?tab=readme-ov-file#migrations) for this change to historically update records.

aricodeine commented:
> Please include a [database migration](https://github.com/brandonp2412/Flexify?tab=readme-ov-file#migrations) for this change to historically update records.

I updated the database.dart file and added below method in it and also bumped the version

```dart
from43To44: (Migrator m, Schema44 schema) async {
          await m.database.customUpdate(
            "UPDATE plans SET exercises = REPLACE(exercises, ',', '~')",
          );
        },
``` 

after running the migrate script, I am getting an error in the below auto-generated file:
<img width="965" height="425" alt="image" src="https://github.com/user-attachments/assets/bdb0bb37-e08f-450b-82f2-a5d4cbbf52a2" />


please let me know if i need to update this file manually to specify int datatype for versions variable explicitly.


Also the migration script, modified a lot of scripts, I will add database.g.dart, schema_v44.dart, two schema.dart files

Anything else needed in the PR?

aricodeine commented:
> Please include a [database migration](https://github.com/brandonp2412/Flexify?tab=readme-ov-file#migrations) for this change to historically update records.

<img width="1257" height="353" alt="image" src="https://github.com/user-attachments/assets/9a61f8ab-cf1e-4ed0-8b86-1a180941e48d" />


@brandonp2412 quick question, some of the data that you can see in above image, are the records for those not present in PlanExercises Table?

Since PlanExercises table has PlanId column, I assumed that relevant data might be present in the table.

I was planning to use PlanExercises table to get a list of exercises that is related to a specific plan Id and migrate the existing exercises column in Plan table to store it separated by '~' instead.


Please let me know if I'm missing something

brandonp2412 commented:
As for your 1. I think the data should all be stored in plan exercises table and right now there will be a mis-match where plan has a string list of the exercises, which we should move away from

aricodeine commented:
Alright so I am thinking of modifying following things:
1) Remove exercises column from the Plans table
2) Update the logic so that only the correct exercise records are created in the PlanExercises Table
3) Retrieve the PlanExercises records where PlanId= Plan.Id where Plan is current plan record displayed in the Start_Plan_Page.dart and display those records one below another (probably we just need to put them in a string separated by comma, which the code already does it)
3) After 1), 2) and 3) are done, then I will work on a final migration step.

Please let me know if you have any thoughts on this.

brandonp2412 commented:
Hey thanks for the work so far, I have a few comments:

1. Some of your code is expecting non-null variables where it's possible for them to end up null:
```dart
            exercises: exercisesMap![plan.id] ?? '',
```
2. Related to 1. when you import plans as a CSV most plan related pages are broken.
3. Please ensure all unit tests pass
```bash
flutter test
```

aricodeine commented:
Sure, will check. Can you provide a sample csv file?

brandonp2412 commented:
Thanks for the contributions boss! I'll take it from here 😎
